### PR TITLE
Search: Toggle Search based on search query

### LINKED ISSUE #23641: Search: open state & query string via url parameter
Currently search does not close when you hit browser back button or click on another page in the sidemenu. 

We should hook up https://github.com/grafana/grafana/blob/master/public/app/features/search/components/SearchWrapper.tsx to redux and move the open state to redux location (url query parameter). So you can open it via ?search=open, that way back browser button will work full page reload will reopen search.

Then we should then consider adding query & filter state to url as well.

### PR DESCRIPTION
**Pull request by Clarity-89**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #23641

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by Clarity-89** [comment]
@torkelo I've added the search query params for opening `DashboardSearch` and it's working fine except for the issue that all React components are re-rendered on location update, so when `DashboardSearch` is opened/closed the components on that page are re-rendered. Normally it's not a big deal, however in case of `ManageDashboards` all the search results are refetched when search is toggled. I wonder if it's an expected behaviour? Wrapping components in `memo` doesn't help since re-renders happen all the way from top. I'm assuming it comes from some angular wrappers.

Additionally all the components that have `navModel` as a prop will re-render despite being wrapped in React.memo, since it only does a shallow comparison of `navModel` by default.

File: public/app/features/search/components/SearchWrapper.tsx [hunk 1]
```diff
@@ -28,7 +54,17 @@ export const SearchWrapper: FC = () => {
...
+  return isOpen ? <DashboardSearch onCloseSearch={closeSearch} payload={payload} /> : null;
+});
+
+const mapStateToProps = (state: StoreState) => {
+  return { search: getLocationQuery(state.location).search };
+};
+
+const mapDispatchToProps = {
+  updateLocation,
 };
```
Comment (torkelo) [inline public/app/features/search/components/SearchWrapper.tsx:68]: Type both of these using the redux types MapStateToProps MapDispatchToProps  
https://github.com/grafana/grafana/blob/master/public/app/features/dashboard/dashgrid/DashboardPanel.tsx#L143

File: public/app/features/search/components/SearchWrapper.tsx [hunk 2]
```diff
@@ -1,23 +1,49 @@
...
   const [payload, setPayload] = useState({});
+  const isOpen = search === 'open';
+
+  const closeSearch = () => {
+    if (search === 'open') {
+      updateLocation({
+        query: {
+          search: 'closed',
+        },
+        partial: true,
```
Comment (torkelo) [inline public/app/features/search/components/SearchWrapper.tsx:27]: this should be search: null  to remove the query parameter

File: public/app/features/search/components/SearchWrapper.tsx [hunk 3]
```diff
@@ -28,7 +54,17 @@ export const SearchWrapper: FC = () => {
       appEvents.off(CoreEvents.showDashSearch, openSearch);
       appEvents.off(CoreEvents.hideDashSearch, closeOnItemClick);
     };
```
Comment (torkelo) [inline public/app/features/search/components/SearchWrapper.tsx:56]: you do not need these events anymore, now anyone can open search by adding the query parameter (see keybindigSrv does this for a lot stuff already top open edit mode or inspect)

File: public/app/features/search/components/SearchWrapper.tsx [hunk 4]
```diff
@@ -1,23 +1,49 @@
...
     const openSearch = (payload: OpenSearchParams) => {
-      setIsOpen(true);
-      setPayload(payload);
+      if (search !== 'open') {
+        setPayload(payload);
+        updateLocation({
+          query: { search: 'open' },
+          partial: true,
+        });
+      }
```
Comment (torkelo) [inline public/app/features/search/components/SearchWrapper.tsx:40]: this can go away, just take open state directly from location.query.search

**Review by torkelo — COMMENTED** [review]
Not sure what you mean what is re-rendered , and when?

**Comment by Clarity-89** [comment]
@torkelo if you go to `/dashboards` and open Dashboard search, then close it, you can see the loading happens again for manage dashboards list.

File: public/app/features/search/components/SearchWrapper.tsx (hunk 3 above)
Comment (Clarity-89) [reply public/app/features/search/components/SearchWrapper.tsx:56]: Some of those events send `payload`, like when you click on dashboard name and the search defaults to current folder, so I think this is necessary until those evens are updated as well.

Comment (torkelo) [reply public/app/features/search/components/SearchWrapper.tsx:62]: yea, should be just adding a folderId= (or what ever the payload is as url parameter

**Comment by torkelo** [comment]
Do they share any state?

**Comment by Clarity-89** [comment]
> Do they share any state?

As far as I can see, it's the root store. I see `store updating angular $location.url` in the console. Just wanted to check if it's a known issue/intended behavior before looking closer into it.

**Comment by Clarity-89** [comment]
@torkelo I checked the code and it looks like components rendered via `react-container` directive in [routes.ts](https://github.com/grafana/grafana/blob/master/public/app/routes/routes.ts#L162) are rendered on location change, which is what happens when the search is toggled. So it seems like toggling search via url change will cause the react component on that page to re-render.

**Comment by torkelo** [comment]
hm.. dammit, we can add reloadOnSearch: false to most routes, but some I might need it to

**Review by alexanderzobnin — CHANGES_REQUESTED** [review]
Works well! Only found a small issue related to folders: 
1. Open dashboard from the folder.
1. Click on the folder name to run search in the current folder.
1. `folder=current` URL param will be added
1. Close search.
1. `folder=current` is still present in url params, so if you open search again by clicking to the dashboard name, `folder:current` will be added to the search filter.

File: public/app/features/search/components/SearchWrapper.tsx [hunk 5]
```diff
@@ -1,34 +1,48 @@
...
+export type Props = OwnProps & DispatchProps;
+
+export const SearchWrapper: FC<Props> = memo(({ search, folder, updateLocation }) => {
+  const isOpen = search === 'open';
+
+  const closeSearch = () => {
+    if (search === 'open') {
+      updateLocation({
+        query: {
+          search: null,
```
Comment (alexanderzobnin) [inline public/app/features/search/components/SearchWrapper.tsx:29]: Probably this should fix a bug
```suggestion
          search: null,
          folder: null,
```

Comment (Clarity-89) [reply public/app/features/search/components/SearchWrapper.tsx:29]: Yeah, good catch, forgot to reset the folder :)

**Review by alexanderzobnin — APPROVED** [review]
LGTM!

**Comment by Clarity-89** [comment]
@torkelo I added `reloadOnSearch: false` to the manage dashboards urls and it works, no more search results re-fetching when toggling search. Not sure which other routes this should be added to.

**Comment by torkelo** [comment]
That is fine for now

**Review by torkelo — APPROVED** [review]

**Pull request merged**